sync3/caches/global.go

```diff
@@ -385,3 +385,22 @@ func (c *GlobalCache) OnNewEvent(
 	}
 	c.roomIDToMetadata[ed.RoomID] = metadata
 }
+
+func (c *GlobalCache) OnInvalidateRoom(ctx context.Context, roomID string) {
+	c.roomIDToMetadataMu.Lock()
+	defer c.roomIDToMetadataMu.Unlock()
+
+	metadata, ok := c.roomIDToMetadata[roomID]
+	if !ok {
+		logger.Warn().Str("room_id", roomID).Msg("OnInvalidateRoom: room not in global cache")
+		return
+	}
+
+	logger.Warn().Any("before", metadata).Msgf("DMR: invalidate GLOBALLLLLL room %s", roomID)
+	err := c.store.ResetMetadataState(metadata)
```

Comment:
Presumably needs to be written back to `c.roomIDToMetadata`?

Reply:
Here `metadata` is a pointer and I'm updating `*metadata` in-place. We have a lock on roomIDToMetadataMu in the caller so this might be safe. I wasn't completely sure though, see e.g. https://github.com/matrix-org/sliding-sync/blob/1b14ac3d1f8b828b378acf854c9ff350339f2895/state/storage.go#L311-L312

Reply:
From discussion elsewhere:

> it should be safe so long as you're doing that from the subscriber goroutine, and not per-conn goroutines